File: services/rag/src/pai_rag_service/retrieval/fusion.py
"""Merge two ranked lists with Reciprocal Rank Fusion.
RRF looks only at rank, so BM25 and cosine, which share no scale, need no implicit weight.
"""

from __future__ import annotations

from dataclasses import dataclass
from enum import StrEnum

__all__ = ["MatchedBy", "RRF_K", "Ranked", "fuse"]

#: RRF constant from Cormack et al. Not a tuning parameter: it flattens the gap between rank 1 and rank 2.
RRF_K = 60.0


class MatchedBy(StrEnum):
    """Why a chunk showed up in the results."""

    KEYWORD = "keyword"
    SEMANTIC = "semantic"
    BOTH = "both"


@dataclass(slots=True)
class Ranked:
    chunk_id: int
    score: float
    matched_by: MatchedBy


def fuse(keyword: list[int], semantic: list[int], limit: int) -> list[Ranked]:
    """Merge two already-ranked lists (best first) into one."""
    merged: list[Ranked] = []
    seen: dict[int, int] = {}

    def contribute(ids: list[int], source: MatchedBy) -> None:
        for index, chunk_id in enumerate(ids):
            # Ranks count from 1: rank 0 would make the denominator `k` and erase the gap to second place.
            contribution = 1.0 / (RRF_K + index + 1)
            at = seen.get(chunk_id)
            if at is None:
                seen[chunk_id] = len(merged)
                merged.append(Ranked(chunk_id, contribution, source))
                continue
            row = merged[at]
            row.score += contribution
            if row.matched_by is not source:
                row.matched_by = MatchedBy.BOTH

    contribute(keyword, MatchedBy.KEYWORD)
    contribute(semantic, MatchedBy.SEMANTIC)

    # Score descending, ties by ascending id, so identical questions always give an identical order.
    merged.sort(key=lambda row: (-row.score, row.chunk_id))
    return merged[:limit]

File: harness/crates/pai-rag-core/src/fusion.rs
use std::{cmp::Ordering, collections::HashMap};

use serde::{Deserialize, Serialize};

/// Reciprocal-rank constant used by the document retriever.
pub const RRF_K: f64 = 60.0;

#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum MatchedBy {
    Keyword,
    Semantic,
    Both,
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
pub struct Ranked {
    pub chunk_id: i64,
    pub score: f64,
    pub matched_by: MatchedBy,
}

/// Merge two best-first lists with deterministic reciprocal-rank fusion.
pub fn fuse(keyword: &[i64], semantic: &[i64], limit: usize) -> Vec<Ranked> {
    let mut merged = Vec::new();
    let mut seen = HashMap::new();

    contribute(&mut merged, &mut seen, keyword, MatchedBy::Keyword);
    contribute(&mut merged, &mut seen, semantic, MatchedBy::Semantic);

    merged.sort_by(|left, right| {
        right
            .score
            .partial_cmp(&left.score)
            .unwrap_or(Ordering::Equal)
            .then_with(|| left.chunk_id.cmp(&right.chunk_id))
    });
    merged.truncate(limit);
    merged
}

fn contribute(
    merged: &mut Vec<Ranked>,
    seen: &mut HashMap<i64, usize>,
    ids: &[i64],
    source: MatchedBy,
) {
    for (index, chunk_id) in ids.iter().copied().enumerate() {
        let contribution = 1.0 / (RRF_K + index as f64 + 1.0);
        if let Some(&at) = seen.get(&chunk_id) {
            let row = &mut merged[at];
            row.score += contribution;
            if row.matched_by != source {
                row.matched_by = MatchedBy::Both;
            }
        } else {
            seen.insert(chunk_id, merged.len());
            merged.push(Ranked {
                chunk_id,
                score: contribution,
                matched_by: source,
            });
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn shared_hits_win_and_ties_are_stable() {
        let ranked = fuse(&[9, 2, 7], &[7, 2, 8], 4);
        assert_eq!(ranked[0].chunk_id, 7);
        assert_eq!(ranked[0].matched_by, MatchedBy::Both);
        assert_eq!(ranked[1].chunk_id, 2);
        assert_eq!(ranked[1].matched_by, MatchedBy::Both);
        assert_eq!(ranked[2].chunk_id, 9);
        assert_eq!(ranked[3].chunk_id, 8);
    }
}
